### ISSUE TITLE
operator always deletes first PodDisruptionBudgets during node drain

### ISSUE THREAD

**Issue by r0bj** [rook/rook#4290]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The Rook operator always deletes first PDB (rook-ceph-osd-0) during node drain.

**Expected behavior:**
The Rook operator deletes correct PDB based on canary pod during node drain.

**How to reproduce it (minimal and precise):**

Relevant workloads:
```
rook-ceph-drain-canary-dev-k8s-storage-r2-685cb494c8-wlsbn   1/1     Running   0          13m     10.205.183.202   dev-k8s-storage-r2   <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r3-79cd664bb9-lqbch   1/1     Running   0          10m     10.205.151.221   dev-k8s-storage-r3   <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r4-58d858cb7b-vprlp   1/1     Running   0          6m50s   10.205.25.8      dev-k8s-storage-r4   <none>           <none>
rook-ceph-osd-0-79d844bf67-hhkk7                             1/1     Running   0          13m     10.205.183.200   dev-k8s-storage-r2   <none>           <none>
rook-ceph-osd-1-78f7df54b-82mrz                              1/1     Running   0          142m    10.205.151.206   dev-k8s-storage-r3   <none>           <none>
rook-ceph-osd-2-8547d68cf9-zdqpl                             1/1     Running   0          140m    10.205.25.42     dev-k8s-storage-r4   <none>           <none>
```

PodDisruptionBudgets:
```
rook-ceph-osd-0     N/A             0                 0                     7m57s
rook-ceph-osd-1     N/A             0                 0                     20m
rook-ceph-osd-2     N/A             0                 0                     20m
```

Draining first node `dev-k8s-storage-r2`:
```
rook-ceph-drain-canary-dev-k8s-storage-r2-685cb494c8-r5stz   0/1     Pending   0          40s    <none>           <none>               <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r3-79cd664bb9-lqbch   1/1     Running   0          14m    10.205.151.221   dev-k8s-storage-r3   <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r4-58d858cb7b-vprlp   1/1     Running   0          10m    10.205.25.8      dev-k8s-storage-r4   <none>           <none>
rook-ceph-osd-0-79d844bf67-lrlv2                             0/1     Pending   0          35s    <none>           <none>               <none>           <none>
rook-ceph-osd-1-78f7df54b-82mrz                              1/1     Running   0          146m   10.205.151.206   dev-k8s-storage-r3   <none>           <none>
rook-ceph-osd-2-8547d68cf9-zdqpl                             1/1     Running   0          143m   10.205.25.42     dev-k8s-storage-r4   <none>           <none>
```

```
rook-ceph-osd-1     N/A             0                 0                     23m
rook-ceph-osd-2     N/A             0                 0                     23m
```

Everything is ok:
1. pod `rook-ceph-drain-canary-dev-k8s-storage-r2-685cb494c8-r5stz` was evicted from node `dev-k8s-storage-r2`
2. rook operator removed PDB `rook-ceph-osd-0`
3. pod `rook-ceph-osd-0-79d844bf67-lrlv2` was evicted from node `dev-k8s-storage-r2` as expected
4. drain finished successfully

Now, let's drain second node.
Draining first node `dev-k8s-storage-r3`:
```
rook-ceph-drain-canary-dev-k8s-storage-r2-685cb494c8-r5stz   1/1     Running   0          5m49s   10.205.183.241   dev-k8s-storage-r2   <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r3-79cd664bb9-mcwvx   0/1     Pending   0          14s     <none>           <none>               <none>           <none>
rook-ceph-drain-canary-dev-k8s-storage-r4-58d858cb7b-vprlp   1/1     Running   0          15m     10.205.25.8      dev-k8s-storage-r4   <none>           <none>
rook-ceph-osd-0-79d844bf67-lrlv2                             1/1     Running   0          5m44s   10.205.183.199   dev-k8s-storage-r2   <none>           <none>
rook-ceph-osd-1-78f7df54b-82mrz                              1/1     Running   0          151m    10.205.151.206   dev-k8s-storage-r3   <none>           <none>
rook-ceph-osd-2-8547d68cf9-zdqpl                             1/1     Running   0          148m    10.205.25.42     dev-k8s-storage-r4   <none>           <none>
```

```
rook-ceph-osd-1     N/A             0                 0                     28m
rook-ceph-osd-2     N/A             0                 0                     28m
```

What happened:
1. pod `rook-ceph-drain-canary-dev-k8s-storage-r3-79cd664bb9-mcwvx` was evicted from node `dev-k8s-storage-r3` as expected
2. for some reason rook operator removed PDB `rook-ceph-osd-0` (node `dev-k8s-storage-r2`) instead of `rook-ceph-osd-1` which is corresponding to pod `rook-ceph-osd-1-78f7df54b-82mrz` on node `dev-k8s-storage-r3`
3. node cannot be drained because PDB wont allow for draining pod `rook-ceph-osd-1-78f7df54b-82mrz`

From my observation, rook operator always remove first PDB whichever node is drained.

**Environment**:
* OS (e.g. from /etc/os-release): ubuntu 18.04
* Kernel (e.g. `uname -a`): 4.15.0-42-generic
* Rook version (use `rook version` inside of a Rook Pod): v1.1.6
* Storage backend version (e.g. for ceph do `ceph -v`): 14.2.4
* Kubernetes version (use `kubectl version`): 1.16.2

**Comment by leseb** [comment]
@rohantmp PTAL

**Comment by rohantmp** [comment]
@r0bj any logs from the rook operator pod with the line clusterdisruption would be useful, also the the output of `kubectl get deployments -l app=rook-ceph-drain-canary`

This could be an issue with stale drain-canary deployments, and if that is the case #4272 should fix this.

**Comment by r0bj** [comment]
Logs from operator during drain:
```
2019-11-09 10:33:36.801243 I | op-cluster: Updated node dev-k8s-storage-r3 is not valid and could not get added to cluster in namespace rook-ceph.
2019-11-09 10:33:37.323218 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:33:51.345285 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:33:52.161255 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:33:52.999278 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:33:59.575386 W | op-mon: mon d not found in quorum, waiting for timeout before failover
2019-11-09 10:34:22.163361 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:34:22.996000 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:34:34.245105 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
2019-11-09 10:34:45.015379 W | op-mon: mon d not found in quorum, waiting for timeout before failover
```

deployments status:
```
rook-ceph-drain-canary-dev-k8s-storage-r2   1/1     1            1           10d
rook-ceph-drain-canary-dev-k8s-storage-r3   0/1     1            0           10d
rook-ceph-drain-canary-dev-k8s-storage-r4   1/1     1            1           10d
```

**Comment by rohantmp** [comment]
Was the deployment status taken when the problem was occuring? It indicates that only one deployment was pending, but 
```
2019-11-09 10:34:34.245105 I | clusterdisruption-controller: pg health: all PGs in cluster are clean. detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]
```
implies that all 3 were pending.

**Comment by r0bj** [comment]
Yes, it was during draining one node `dev-k8s-storage-r3`. Not sure why it says `detected drains on hosts: [dev-k8s-storage-r2 dev-k8s-storage-r3 dev-k8s-storage-r4]`.

**Comment by rohantmp** [comment]
Oh, were these PVC based OSDs with `portable: true`?

**Comment by r0bj** [comment]
I'm using host-based cluster, this is cluster definition:
```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  cephVersion:
    image: ceph/ceph:v14.2.4-20190917
    allowUnsupported: false
  dataDirHostPath: /data/rook
  skipUpgradeChecks: false
  mon:
    count: 3
    allowMultiplePerNode: false
  mgr:
    modules:
    - name: pg_autoscaler
      enabled: true
  dashboard:
    enabled: true
    ssl: false
  monitoring:
    enabled: false
    rulesNamespace: rook-ceph
  network:
    hostNetwork: false
  rbdMirroring:
    workers: 0
  placement:
    all:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: k8s-role
              operator: In
              values:
              - k8s-storage
      tolerations:
      - key: node-role.kubernetes.io/storage
        operator: Exists
  resources:
    mgr:
      limits:
        memory: 1Gi
      requests:
        cpu: 20m
        memory: 300Mi
    mon:
      limits:
        memory: 8Gi
      requests:
        cpu: 20m
        memory: 800Mi
    osd:
      requests:
        cpu: 20m
        memory: 2Gi
  storage:
    useAllNodes: true
    useAllDevices: false
    directories:
    - path: /data/rook
  disruptionManagement:
    managePodBudgets: true
    osdMaintenanceTimeout: 30
```

Requested command output:
```
# ceph osd tree
ID CLASS WEIGHT  TYPE NAME                   STATUS REWEIGHT PRI-AFF
-1       4.39229 root default
-2       1.46410     host dev-k8s-storage-r2
 0   hdd 1.46410         osd.0                   up  1.00000 1.00000
-3       1.46410     host dev-k8s-storage-r3
 1   hdd 1.46410         osd.1                   up  1.00000 1.00000
-7       1.46410     host dev-k8s-storage-r4
 2   hdd 1.46410         osd.2                   up  1.00000 1.00000
```

**Comment by r0bj** [comment]
It seems that it's fixed in rook 1.1.7.

**Comment by rohantmp** [comment]
Yeah, you're issue helped me find the cause! Sorry I forgot to update the issue after sending the patch!
